Initialize Hound package / module

grahamc opened the pull request:
###### Motivation for this change

https://search.nix.gsc.io is powered by Hound, and I figured it couldn't hurt to upstream the code for it.
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

grahamc committed:
hound: init at 20160919

grahamc committed:
hound: init module

joachifm commented on nixos/modules/misc/ids.nix:
```diff
@@ -276,6 +276,7 @@
       telegraf = 256;
       gitlab-runner = 257;
       postgrey = 258;
+      hound = 259;
```
What sort of data does `hound` generate? Is it not sufficient to ensure eventually consistent ownership via `preStart` or similar mechanisms?

grahamc replied:
I'm really not sold on using preStart's to chmod a bunch of data. One of my hound installs has 700GB of owned data. I definitely don't want to be running chmod during a restart.

Mic92 commented on pkgs/development/tools/misc/hound/default.nix:
```diff
@@ -0,0 +1,28 @@
+
+  goDeps = ./deps.nix;
+
+  # TODO: add metadata https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes
```
obsolete.

Mic92 commented on nixos/modules/services/search/hound.nix:
```diff
@@ -0,0 +1,119 @@
+        '';
+      };
+
+      extraGroups = mkOption {
```
what is the use case for this option?

Mic92 commented:
Last but not least the test should be also added to `release.nix`:

```patch
diff --git a/nixos/release.nix b/nixos/release.nix
index 10c624a..fbd3efd 100644
--- a/nixos/release.nix
+++ b/nixos/release.nix
@@ -245,6 +245,7 @@ in rec {
   tests.gnome3-gdm = callTest tests/gnome3-gdm.nix {};
   tests.grsecurity = callTest tests/grsecurity.nix {};
   tests.hibernate = callTest tests/hibernate.nix {};
+  tests.hound = callTest tests/hound.nix {};
   tests.i3wm = callTest tests/i3wm.nix {};
   tests.installer = callSubTests tests/installer.nix {};
   tests.influxdb = callTest tests/influxdb.nix {};
```

grahamc commented:
@Mic92 I addressed the feedback, other than adding it to the release. I don't really want it to block the tested job. Will this?

Mic92 commented:
@grahamc well, otherwise this tests is never evaluated by hydra. What do you mean by block?

grahamc commented:
Fixed. :) Thanks for the review!

ttuegel commented:
This breaks `nixos-rebuild` with
```
Option `services.hound.config' has no description.
```
even for configurations that don't have Hound enabled.

grahamc commented:
D'oh, thank you for the heads up. Curious that it errors there, but didn't error in my local usage.  Preparing a fix.

grahamc commented:
Fixed up. Sorry about that. Is this a new check not in 16.09? I've used this module in 16.09.

bjornfor commented:
```
$ nix-env -f . -qa '*' --meta --xml --drv-path --show-trace
[...]
error: while querying the derivation named ‘go1.7-hound-unstable-20160919-f95e9a9’:
attribute ‘homepage’ missing, at /home/bfo/proj/code/forks/nixpkgs3/pkgs/development/tools/misc/hound/default.nix:19:28
```

grahamc commented:
@bjornfor guh, sorry. :flushed: 

I fixed hound in https://github.com/NixOS/nixpkgs/commit/104d6965af3254a2fc6777de847abd1c7b4af237

and in testing found another PR broke it as well, which I fixed in https://github.com/NixOS/nixpkgs/commit/33ac1e1d63f6f2f4f6b997a8688233102867acd2 (/cc @Profpatsch, @Mic92 to learn what I just learned, too :) )

is this the best way to find these sorts of errors in the future?

This seems like a _really_ good test to have in Travis-CI -- which _shouldn't_ be so hard to run as to fail regularly.

bjornfor commented:
I'm surprised travis doesn't run that check. I have put that command in a script, for my own use. It catches a lot of those eval issues:

```
$ cat ~/bin/nix-check-before-push.sh 
#!/bin/sh
# This checks for evaluation errors in nixpkgs (and nixos?) git repositories.
# Run it before commit/push.

set -x
nix-env -f . -qa \* --meta --xml --drv-path --show-trace >/dev/null || { echo FAILED; exit 1; }
nix-build pkgs/top-level/release.nix -A tarball || { echo FAILED; exit 1; }
```
